triton-constraint-circuit/src/lib.rs

```diff
@@ -1471,4 +1705,316 @@ mod tests {
+    ///          C* --- degree-lowering ----> C* || k
+    /// ```
+    ///
+    /// The point of this elaboration is that in this particular case, testing completeness
+    /// and soundness require the same code path. If `input` and `input || some_expr(input)`
+    /// work for circuits before and after degree lowering, this fact establishes
+    /// both completeness and soundness simultaneously.
+    ///
+    /// Shrinking on this test is disabled because we noticed some weird ass behavior.
+    /// In short, shrinking does not play ball with the arbitrary circuit generator;
+    /// it seems to make the generated circuits *more* complex, not less so.
+    #[proptest(cases = 1000, max_shrink_iters = 0)]
+    fn node_substitution_is_complete_and_sound(
+        #[strategy(arbitrary_circuit_monad(10, 10, 10, 200, 10))] mut multicircuit_monad: Vec<
+            ConstraintCircuitMonad<SingleRowIndicator>,
+        >,
+        #[strategy(vec(arb(), ConstraintCircuitMonad::num_main_inputs(&#multicircuit_monad)))]
+        #[filter(!#main_input.is_empty())]
+        main_input: Vec<BFieldElement>,
+        #[strategy(vec(arb(), ConstraintCircuitMonad::num_aux_inputs(&#multicircuit_monad)))]
+        #[filter(!#aux_input.is_empty())]
+        aux_input: Vec<XFieldElement>,
+        #[strategy(vec(arb(), ConstraintCircuitMonad::num_challenges(&#multicircuit_monad)))]
+        challenges: Vec<XFieldElement>,
+        #[strategy(arb())] substitution_node_index: usize,
+    ) {
+        let mut main_input = Array2::from_shape_vec((1, main_input.len()), main_input).unwrap();
+        let mut aux_input = Array2::from_shape_vec((1, aux_input.len()), aux_input).unwrap();
+
+        let output_before_lowering = multicircuit_monad
+            .iter()
+            .map(|constraint| {
+                constraint.circuit.borrow().evaluate(
+                    main_input.view(),
+                    aux_input.view(),
+                    &challenges,
+                )
+            })
+            .collect_vec();
+
+        // apply one step of degree-lowering
+        let num_nodes = multicircuit_monad[0].builder.all_nodes.borrow().len();
+        let (mut substitution_node_id, mut substituted) = multicircuit_monad[0]
+            .builder
+            .all_nodes
+            .borrow()
+            .iter()
+            .nth(substitution_node_index % num_nodes)
+            .map(|(i, n)| (*i, n.clone()))
+            .unwrap();
+        let mut j = 0;
+        while substituted.circuit.borrow().is_zero() && j < num_nodes {
+            (substitution_node_id, substituted) = multicircuit_monad[0]
+                .builder
+                .all_nodes
+                .borrow()
+                .iter()
+                .nth((substitution_node_index + j) % num_nodes)
+                .map(|(i, n)| (*i, n.clone()))
+                .unwrap();
+            j += 1;
+        }
+        prop_assert_ne!(j, num_nodes, "no suitable nodes to substitute");
```

Comment:
The following removes the duplicated code. Since `substituted` is no longer mutable, the compiler correctly detects that it is never used.

```suggestion
        let (&substitution_node_id, _) = multicircuit_monad[0]
            .builder
            .all_nodes
            .borrow()
            .iter()
            .cycle()
            // `% num_nodes` is not even necessary due to the `cycle()`
            // but I think it communicates intent. Happy with either.
            // This comment is only intended for the review, not the code.
            .skip(substitution_node_index % num_nodes)
            .take(num_nodes)
            .find(|(_, monad)| !monad.circuit.borrow().is_zero())
            .expect("no suitable nodes to substitute");
```

Reply:
Further improved with `find_map`.